File: decelium_wallet/networkchannels/http_client.js
/* disable-eslint-disable disable-import/first */

import fetch from 'cross-fetch';
import * as ipfsClient from 'ipfs-http-client';

let fs = undefined;
let path = undefined;
let base58btc= undefined;
let encode= undefined;
let CID=  undefined;

class http_client_wrapped {
    constructor(url_version = null, api_key = null, port = null) {
        this.url_version = url_version;
        this.api_key = api_key;
        this.port = port;
    }
    
   async applyAlternateProcessing(filter, sourceId) {
        let pathVar;
        let fsVar; // Hiding from webpack's idiotic module scanner
       
        if (fs == undefined)
        {
            path = await import('path');
            fs = await import('fs');
        }
       if (sourceId === "create_ipfs" && filter.file_type === undefined) {
           filter.file_type = 'ipfs';
       }

       // Check if payload_type is not defined and payload is a string, then set payload_type to 'local_path'
       if (sourceId === "create_ipfs" && filter.payload_type === undefined && typeof filter.payload === 'string') {
           filter.payload_type = 'local_path';
       }

       // Check if payload_type is not defined and payload is an object (note: null is also considered an object in JavaScript)
       if (sourceId === "create_ipfs" && filter.payload_type === undefined && typeof filter.payload === 'object' && filter.payload !== null) {
           filter.payload_type = 'raw_file_list';
       }

        if (sourceId === "create_ipfs" && filter['file_type'] === 'ipfs' )
        {
            if (filter['payload_type'] != 'local_path' && filter['payload_type'] != 'raw_file_list')
                return {error:"only payload_type local_path and raw_file_list is accepted with create_ipfs"}
        }
        if (sourceId === "create_ipfs"  && 'file_type' in filter && filter['file_type'] === 'ipfs' && 'payload_type' in filter) {
            try {
                let itemsToAdd = [];
                if (filter['payload_type'] === 'local_path')
                    itemsToAdd = await this.loadFileData(filter['payload']);
                if (filter['payload_type'] === 'raw_file_list')
                    itemsToAdd =filter['payload'];
                    
                let connection_settings = filter.connection_settings;
                if (connection_settings == undefined) {
                    const url = new URL(this.url_version);
                    connection_settings = {
                        host: url.hostname,
                        protocol: url.protocol.slice(0, -1),
                        port: 5002
                    }
                }
                
                if (connection_settings.host == undefined) return {"error":"ipfs host must be specified in connection_settings"};
                if (connection_settings.port == undefined) return {"error":"ipfs port must be specified in connection_settings"};
                if (connection_settings.protocol == undefined) return {"error":"ipfs protocol must be specified in connection_settings"};
                if (connection_settings.headers == undefined)
                    connection_settings.headers = {};
                console.log("INSPECTING connection_settings");
                console.log(connection_settings);
                this.ipfs = ipfsClient.create(connection_settings);
                let generator = await this.ipfs.addAll(itemsToAdd, { wrapWithDirectory: true });
                let addedItems = [];

                for await (const item of generator) {
                    // For each added item, to avoid errors, painstakingly verify each file.
                    const pinResult = await this.ipfs.pin.add(item.cid);
                    let isPinned = false;
                    for await (const pin of this.ipfs.pin.ls({ paths: item.cid })) {
                        if (pin.cid.toString() === item.cid.toString()) {
                            isPinned = true;
                            break;
                        }
                    }
                    
                    if(isPinned == false)
                    {
                        console.error('Error Could not add pin to your pinning service:'+ item.cid.toString());
                        throw new Error("Could not add pin to pinning service.");                    
                    }
                    item.cid = item.cid.toString();
                    item.name = item.path.toString();
                    if (item.name == "")
                    {
                        item.root = true;
                    }
                    addedItems.push(item);

                }

                return addedItems;
            } catch (error) {
                console.error('Error adding to IPFS:', error);
                return { error: 'Error adding to IPFS:' + error.toString() }

            }
            
        } else {
            return undefined;
        }
    }
    
    __run_query(q, remote = true, show_url = false) {
        let wait_seconds = 120; 
        let re_query_delay = 5;
        if (!('api_key' in q) || q['api_key'] === null) {
            q['api_key'] = this.api_key;
        }
        return this.query(q, this.current_attr, {remote, url_version: this.url_version, wait_seconds, re_query_delay, show_url});
    }
    
    async loadFileData (path_in) {
        const getFilesRecursive = (dir) => {
            let results = [];
            const list = fs.readdirSync(dir);
            list.forEach((file) => {
                const filePath = path.join(dir, file);
                const stat = fs.statSync(filePath);
                if (stat && stat.isDirectory()) {
                    results = results.concat(getFilesRecursive(filePath));
                } else {
                    results.push(filePath);
                }
            });
            return results;
        };

        let itemsToAdd = [];
        if (fs.statSync(path_in).isDirectory()) {
            //itemsToAdd.push({
            //    path: 'root',
            //    content: Buffer.from('')  // Representing empty content for root dir
            //});            
            const allFiles = getFilesRecursive(path_in);
            for (const filePath of allFiles) {
                const relativePath = path.relative(path_in, filePath);
                const fileContent = fs.readFileSync(filePath);
                itemsToAdd.push({
                    path: relativePath,
                    content: fileContent
                });
            }
        } else {
            const fileContent = fs.readFileSync(path_in);
            itemsToAdd.push({
                "path": this.getBasename(path_in),
                "content": fileContent,
                "root":true
            });
        }    
        return itemsToAdd;
    }


    async convertFilesToDicts(files) {
        if(base58btc == undefined)
        {
            let base58Module = await import('multiformats/bases/base58');
            let digestModule = await import('multiformats/hashes/digest');
            let cidModule = await import('multiformats/cid');

            base58btc = base58Module.base58btc;
            encode = digestModule.encode;
            CID = cidModule.CID;
        }
        return files.map(file => {
            const digestArray = Object.values(file.cid.multihash.digest);
            const digestBytes = Uint8Array.from(digestArray);
            const mh = encode('sha2-256', digestBytes);
            const cid = CID.create(0, 112, mh);

            const cidString = cid.toString(base58btc);

            return {
                path: file.path,
                cid: cidString
            };
        });
    }





    getBasename(path) {
        return path.split('/').pop();
    } 
    
    async query(filter, source_id, {remote = false, url_version = 'dev', wait_seconds = 120, re_query_delay = 5, show_url = false}) {

        const time_start = Date.now();
        let resp = undefined;
        resp = await this.applyAlternateProcessing(filter, source_id);
        if (resp!== undefined)
        {
            return resp;
        }
        if(show_url === true)
        {
            console.log("THE RAW  QUERY()");
            console.log({source_id,url_version,filter});
        }
        while ((Date.now() - time_start) / 1000 < wait_seconds) {
            resp = await this.query_wait(filter, source_id, {remote, url_version, show_url});
            if (resp && typeof resp === 'object' && 'state' in resp && resp['state'] === 'updating') {

                await new Promise(resolve => setTimeout(resolve, re_query_delay * 1000));
                if ('force_refresh' in filter) {
                    delete filter['force_refresh'];
                }
            } else {
                break;
            }
        }
        if(show_url === true)
        {
            console.log("RETURN FROM QUERY()");
            console.log(resp);
        }
        return resp;
    }

    async query_wait(filter, source_id, {remote = false, url_version = 'dev', show_url = false}) {
        
        if ('__encoded_query' in filter) {
            let dic = this.do_decode(filter['__encoded_query']);
            filter = {...filter, ...dic};
            delete filter['__encoded_query'];
        }

        if ([true, 'http', 'https'].includes(remote)) {
            let resp = await this.query_remote(source_id, filter, url_version, show_url);
            
            return resp;
        }
        return null;
    }

    async query_remote(source_id, query, url_version = 'dev', show_url = false) {

        let data = {};
        data['qtype'] = source_id;
        data['__str_encoded_query'] = this.do_encode_string(query);

        if (show_url) {
            console.log("show_url QUERY REMOTE");
            console.log(url_version);
            console.log(query);
            console.log(data);
        }
        let r = await fetch(url_version, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(data)
        });
        
        let rData = "could not load";
        let dat;
        let responseData;
        let contentType = r.headers.get("content-type");        
        try {

            let rawData = await r.arrayBuffer();
            const decoder = new TextDecoder();       
            const vall = decoder.decode(rawData);
            if (contentType && contentType.includes("application/json")) {
                // Handle JSON response
                responseData = new TextDecoder().decode(rawData);
            } else if (contentType && contentType.includes("text")) {
                // Handle text response
                responseData = new TextDecoder().decode(rawData);
            } else {
                try {
                    // Try to parse as JSON
                    responseData = JSON.parse(new TextDecoder().decode(rawData));
                } catch (e) {
                    try {
                        // If JSON parsing fails, try reading as text
                        responseData = new TextDecoder().decode(rawData);
                    } catch (e) {
                        try {
                            // If text parsing fails, handle it as a Blob
                            responseData = new Blob([rawData]);
                        } catch (e) {
                            console.log("An error occurred while converting the response:", e);
                            responseData = null;
                        }
                    }
                }
            }
        } catch (e) {
            console.log("An error occurred while reading the response:", e);
            responseData = null;
        }        

        try 
        {
            dat = jsondateencode_local.loads(responseData);
        } catch (e) {
            console.log("decelium warning -- could not json decode:");
            console.log(responseData);
            console.log(e);
            dat = rData;
        }        
            
        if (show_url) {
            console.log("show_url return val:");
            console.log(dat);
        }
        
        return dat;
        
    }

    do_encode(request_obj) {
        return false;
    }

    do_decode(data_packet) {
        return false;
    }

    do_encode_string(obj) {
        let string = JSON.stringify(obj);
        let encoded = btoa(string);
        return encoded;
    }

    do_decode_string(data_packet) {
        let user_data_dev = atob(data_packet);
        let data2 = JSON.parse(user_data_dev);
        return data2;
    }


}


    
const proxyHandler = {
    get: function(target, propKey) {
        if (typeof target[propKey] === 'function') {
            return target[propKey];
        } else {
            target.current_attr = propKey;
            return target.__run_query.bind(target);
        }
    }
};

class jsondateencode_local {
    /*
    static loads(dic) {

        let p;
        if (typeof dic === 'string') {
            if (this.isJsonString(dic)) {
                p = JSON.parse(dic, this.datetime_parser);
            } else if (!isNaN(dic)) {
                p = Number(dic);
            } else {
                p = dic;
            }
        } else if (typeof dic === 'object') {
            p = this.datetime_parser('', dic);
        }
        else
        {
            p = dic;
        }

        return p;
    }*/
    static loads(dic) {
        let p;
        if (typeof dic === 'string') {
            if (this.isJsonString(dic)) {
                // Use an arrow function to preserve the context of `this`
                p = JSON.parse(dic, (key, val) => this.datetime_parser(key, val));
            } else if (!isNaN(dic)) {
                p = Number(dic);
            } else {
                p = dic;
            }
        } else if (typeof dic === 'object') {
            p = this.datetime_parser('', dic);
        }
        else
        {
            p = dic;
        }

        return p;
    }    

    static isJsonString(str) {
        try {
            JSON.parse(str);
        } catch (e) {
            return false;
        }
        return true;
    }

    static dumps(dic) {
        return JSON.stringify(dic, this.datedefault);
    }

    static datedefault(key, val) {
        if (Array.isArray(val) && val[0] === '__ref') {
            return ['__ref', ...val.slice(1)];
        }
        if (val instanceof Date) {
            return val.toISOString();
        }
        return val;
    }

    static datetime_parser(key, val) {
        if (typeof val === 'string' && val.includes('T')) {
            const date = new Date(val);
            if (!isNaN(date)) {
                return date;
            }
        } else if (typeof val === 'object') {
            for (let k in val) {
                val[k] = this.datetime_parser(k, val[k]);
            }
        }
        return val;
    }
}


function http_client(url_version = null, api_key = null, port = null) {
    const client = new http_client_wrapped(url_version, api_key, port);
    return new Proxy(client, proxyHandler);
}

export default http_client;
export {http_client};

File: decelium_wallet/networkchannels/http_client.py
import json
import datetime
import pickle
import base64
import requests
import time
import ipfshttpclient
import os

class jsondateencode_local:
    def loads(dic):
        return json.loads(dic,object_hook=jsondateencode_local.datetime_parser)
    def dumps(dic):
        return json.dumps(dic,default=jsondateencode_local.datedefault)

    def datedefault(o):
        if isinstance(o, tuple):
            l = ['__ref']
            l = l + o
            return l
        if isinstance(o, (datetime.date, datetime.datetime,)):
            return o.isoformat()

    def datetime_parser(dct):
        DATE_FORMAT = '%Y-%m-%dT%H:%M:%S'
        for k, v in dct.items():
            if isinstance(v, str) and "T" in v:
                try:
                    dct[k] = datetime.datetime.strptime(v, DATE_FORMAT)
                except:
                    pass
        return dct

class http_client():
    def __init__(self, url_version=None, api_key=None,port=None):
        self.url_version = url_version
        self.api_key = api_key
        self.port = port

    def __getattr__(self, attr):
        self.current_attr = attr
        return self.__run_query

    def __run_query(self, q, remote=True, wait_seconds=120, re_query_delay=5, show_url=False):        
        if 'api_key' not in q or q['api_key'] == None:
            q['api_key'] = self.api_key            
        return self.query(q, self.current_attr, remote=remote, url_version=self.url_version, 
                          wait_seconds=wait_seconds, re_query_delay=re_query_delay, show_url=show_url)

    def create_ipfs_connection(self,connection_settings):
        api = None
        if 'host' not in connection_settings:
            return  {"error": "ipfs host must be specified in connection_settings"},None
        if 'port' not in connection_settings:
            return {"error": "ipfs port must be specified in connection_settings"},None
        if 'protocol' not in connection_settings:
            return {"error": "ipfs protocol must be specified in connection_settings"},None

        # If 'headers' is not present, initialize it to an empty dictionary 
        if 'headers' not in connection_settings:
            connection_settings['headers'] = {}
            
        ipfs_string = "/dns/"+str(connection_settings["host"])+"/tcp/"+str(connection_settings["port"])+"/"+str(connection_settings["protocol"])
            
        if len(connection_settings['headers']) > 0:
            api = ipfshttpclient.connect(ipfs_string,headers=connection_settings['headers'])
        else:
            api = ipfshttpclient.connect(ipfs_string)
        return {"message":"success"}, api

    def apply_alternate_processing(self,filter,source_id):
        if source_id == "create_ipfs":
            return self.processing_create_ipfs(filter,source_id)
        if source_id == "remove_ipfs":
            return self.processing_remove_ipfs(filter,source_id)
        if source_id == "check_pin_status":
            return self.processing_check_pin_status(filter,source_id)
            
        else:
            return None
    def latest_remote_pin_list(self,connection_settings):
        do_refresh = False
        try:
            assert len(self.saved_pins) > 0
        except:
            do_refresh = True
        if do_refresh == True:
            message,api = self.create_ipfs_connection(connection_settings)
            pin_response = api.pin.ls(type='recursive')
            pins = []
            for pin in pin_response['Keys']:
                pins.append(pin)
            self.saved_pins = pins
            
        return self.saved_pins

    def processing_check_pin_status(self, filter,source_id):
        if 'connection_settings' not in filter:
            return {"error": "connection_settings argument required i.e. {host:str, port:int, protocol:str, headers:{authorization:str}}"}
        connection_settings = filter['connection_settings']

        if 'cid' not in filter:
            return {"error":"Need need a 'cid' argument"}
        cid = filter['cid']
        
        pins = self.latest_remote_pin_list(connection_settings)

        if cid in pins:
            return True
        return False

    def processing_create_ipfs(self,filter,source_id):

        if source_id == "create_ipfs" and 'file_type' in filter and filter['file_type'] == 'ipfs' and 'payload_type' in filter:
            # Check if 'connection_settings' is present in the 'filter' dictionary 
            if 'connection_settings' not in filter:
                return {"error": "connection_settings argument required i.e. {host:str, port:int, protocol:str, headers:{authorization:str}}"}
            connection_settings = filter['connection_settings']
            message,api = self.create_ipfs_connection(connection_settings)
            if 'error' in message.keys():
                return message

            if filter['payload_type'] == 'ipfs_pin_list':
                try:
                    # Create an empty directory on IPFS
                    res = api.object.new(template="unixfs-dir")
                    if os.path.isfile( filter['payload']):
                        with open(filter['payload'],'r') as f:
                            filter['payload'] = json.loads(f.read())
                    if 'Links' in filter['payload']:
                        filter['payload'] = filter['payload']['Links']
                    # Iterate over the payload to link each object to the newly created directory
                    for item in filter['payload']:
                        name = item['Name']
                        hash = item['Hash']
                        # Construct the path with the name for the item to be linked under the directory
                        path = f"{res}/{name}"
                        # Use object patch add-link to link the object to the directory
                        res = api.object.patch.add_link(res['Hash'], name, hash)
                        # res = res['Hash']
                        print(res.keys())
                    # After linking all items, the 'empty_dir_cid' will be the CID of the final directory
                    # return  [{'name': res['Name'], 'cid': res['Hash'], 'size': res['Size']}]

                    return  [{'name': '', 'cid': res['Hash'], 'size': None}]
                except Exception as e:
                    import traceback as tb
                    print(tb.format_exc())
                    return {"error": str(e) + tb.format_exc()}

            # Existing condition for 'local_path' payload type 
            elif filter['payload_type'] == 'local_path':
                #res = api.add(filter['payload'],recursive=True)
                original_directory = os.getcwd()
                try:
                    path = filter['payload']
                    if os.path.isdir(path):
                        # If the path is a directory, change the working directory and add recursively 
                        os.chdir(path)
                        res = api.add('.', recursive=True)
                    elif os.path.isfile(path):
                        # If the path is a file, add the file directly without changing the working directory
                        res = api.add(path)
                    else:
                        print("The provided path does not exist or is not a file/directory.")

                finally:
                    os.chdir(original_directory)            
            
                try:
                    dict_list = [{'name': res['Name'], 'cid': res['Hash'], 'size': res['Size']}]
                except:
                    dict_list = [{'name': item['Name'], 'cid': item['Hash'], 'size': item['Size']} for item in res]
                for pin in dict_list:
                    pin_resp = api.pin.add(pin['cid'])
                    #print(pin_resp)

                root = {}
                for item in dict_list:
                    if item['name'] == "":
                        item['root'] = True
                        break
                return dict_list
        return None

    # TODO Migrate updates to JS
    # TODO Migrate FROM js as well
    def processing_remove_ipfs(self,filter,source_id):

        if source_id == "remove_ipfs"  and 'payload_type' in filter and filter['payload_type'] ==  'cid':
            # Check if 'connection_settings' is present in the 'filter' dictionary 
            if 'connection_settings' not in filter:
                return {"error": "connection_settings argument required i.e. {host:str, port:int, protocol:str, headers:{authorization:str}}"}
            connection_settings = filter['connection_settings']
            message,api = self.create_ipfs_connection(connection_settings)
            if 'error' in message.keys():
                return message
            if type(filter['payload']) == str:
                lst = [filter['payload']]
            elif type(filter['payload']) == list: 
                lst = filter['payload']
            else:
                return {"error":"payload must be a CID or list of CIDs"}
            res = {}
            for cid in lst:
                try:
                    resp = api.pin.rm(cid)
                    if cid==resp['Pins'][0]:
                        res[cid]=  {"removed":True}
                    else:
                        res[cid]= {"removed":False,"message":str(resp)}
                except:
                    import traceback as tb
                    res[cid]= {"removed":False,"message":tb.format_exc()}
                    
            return res
        return None

    def query(self, filter, source_id, remote=False, url_version='dev',  wait_seconds = 120, re_query_delay=5, show_url=False):
        time_start = datetime.datetime.utcnow()
        filter_result = self.apply_alternate_processing(filter,source_id)
        if filter_result != None:
            return filter_result        
        
        while (datetime.datetime.utcnow() - time_start).total_seconds() < wait_seconds:
            resp = self.query_wait(filter, source_id, remote, url_version, show_url)
            if type(resp) == dict and 'state' in resp and resp['state'] == 'updating':
                print('.',end='')
                time.sleep(re_query_delay)
                if 'force_refresh' in filter:
                    del(filter['force_refresh'])
            else:
                break
        return resp   

    def query_wait(self, filter, source_id, remote=False, url_version='dev', show_url=False):
        if '__encoded_query' in filter:
            dic = self.do_decode(filter['__encoded_query'])
            filter.update(dic)
            del(filter['__encoded_query'])

        if remote in [True, 'http', 'https']:
            resp = self.query_remote(source_id, filter, url_version, show_url)
            return resp
        return {"error":"No query processed"}

    def query_remote(self, source_id, query, url_version='dev', show_url=False):
        data = {}
        data['qtype'] = source_id
        data['__encoded_query'] = self.do_encode(query)

        if show_url:
            print(url_version)
            print(query)

        r = requests.post(url_version, data=data)

        try:
            dat = jsondateencode_local.loads(r.text)
        except Exception as e:
            print(e)
            dat = r.text
        return dat

    def do_encode(self, request_obj):
        serial = pickle.dumps(request_obj)
        user_data_enc = base64.b64encode(serial).decode("ascii")                   
        return user_data_enc
    
    def do_decode(self, data_packet):
        from urllib.parse import unquote
        user_data_dev = base64.b64decode(unquote(data_packet))   
        data2 = pickle.loads(user_data_dev)
        return data2
    
    def do_encode_string(self, obj):
        string = json.dumps(obj, separators=(',', ':'))
        encoded = base64.b64encode(string.encode('ascii'))
        return encoded.decode('ascii') 

    def do_decode_string(self, data_packet):
        user_data_dev = base64.b64decode(data_packet)                   
        data2 = json.loads(user_data_dev.decode("ascii"))
        return data2
